### DISCUSSION TITLE
How to handle "Resouce 'X/Y/Z' not found in (...)/Zend/Acl.php

### DISCUSSION THREAD

**Discussion by alexh-swdev** [OpenMage/magento-lts#3468] [Questions & Answers]
Hi,

since a while I am seeing the Exception mentioned in the title in my exception.log when I login to the admin area. But I have no real idea how to takle this one.
A concrete example:

>
> 023-08-25T08:03:59+00:00 ERR (3): 
> Zend_Acl_Exception: Resource 'admin/rma/settings/action' not found in /app/vendor/shardj/zf1-future/library/Zend/Acl.php:365
> Stack trace:
> #0 /app/vendor/shardj/zf1-future/library/Zend/Acl.php(641): Zend_Acl->get()
> #1 /app/vendor/shardj/zf1-future/library/Zend/Acl.php(519): Zend_Acl->setRule()
> #2 /app/app/code/core/Mage/Admin/Model/Resource/Acl.php(132): Zend_Acl->deny()
> #3 /app/app/code/core/Mage/Admin/Model/Resource/Acl.php(70): Mage_Admin_Model_Resource_Acl->loadRules()
> #4 /app/app/code/core/Mage/Admin/Model/Session.php(162): Mage_Admin_Model_Resource_Acl->loadAcl()
> #5 /app/app/code/core/Mage/Admin/Model/Observer.php(62): Mage_Admin_Model_Session->login()
> #6 /app/app/code/core/Mage/Core/Model/App.php(1441): Mage_Admin_Model_Observer->actionPreDispatchAdmin()
> #7 /app/app/code/core/Mage/Core/Model/App.php(1419): Mage_Core_Model_App->_callObserverMethod()
> #8 /app/app/Mage.php(509): Mage_Core_Model_App->dispatchEvent()
> #9 /app/app/code/core/Mage/Core/Controller/Varien/Action.php(527): Mage::dispatchEvent()
> #10 /app/app/code/core/Mage/Adminhtml/Controller/Action.php(173): Mage_Core_Controller_Varien_Action->preDispatch()
> #11 /app/app/code/core/Mage/Core/Controller/Varien/Action.php(410): Mage_Adminhtml_Controller_Action->preDispatch()
> #12 /app/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(255): Mage_Core_Controller_Varien_Action->dispatch()
> #13 /app/app/code/core/Mage/Core/Controller/Varien/Front.php(181): Mage_Core_Controller_Varien_Router_Standard->match()
> #14 /app/app/code/core/Mage/Core/Model/App.php(358): Mage_Core_Controller_Varien_Front->dispatch()
> #15 /app/app/Mage.php(743): Mage_Core_Model_App->run()
> #16 /app/index.php(71): Mage::run()
> #17 {main}
> 2023-08-25T08:03:59+00:00 ERR (3): 
> Zend_Acl_Exception: Resource 'admin/Symmetrics_CashTicket/action' not found in /app/vendor/shardj/zf1-future/library/Zend/Acl.php:365
> Stack trace:
> #0 /app/vendor/shardj/zf1-future/library/Zend/Acl.php(641): Zend_Acl->get()
> #1 /app/vendor/shardj/zf1-future/library/Zend/Acl.php(519): Zend_Acl->setRule()
> #2 /app/app/code/core/Mage/Admin/Model/Resource/Acl.php(132): Zend_Acl->deny()
> #3 /app/app/code/core/Mage/Admin/Model/Resource/Acl.php(70): Mage_Admin_Model_Resource_Acl->loadRules()
> #4 /app/app/code/core/Mage/Admin/Model/Session.php(162): Mage_Admin_Model_Resource_Acl->loadAcl()
> #5 /app/app/code/core/Mage/Admin/Model/Observer.php(62): Mage_Admin_Model_Session->login()
> #6 /app/app/code/core/Mage/Core/Model/App.php(1441): Mage_Admin_Model_Observer->actionPreDispatchAdmin()
> #7 /app/app/code/core/Mage/Core/Model/App.php(1419): Mage_Core_Model_App->_callObserverMethod()
> #8 /app/app/Mage.php(509): Mage_Core_Model_App->dispatchEvent()
> #9 /app/app/code/core/Mage/Core/Controller/Varien/Action.php(527): Mage::dispatchEvent()
> #10 /app/app/code/core/Mage/Adminhtml/Controller/Action.php(173): Mage_Core_Controller_Varien_Action->preDispatch()
> #11 /app/app/code/core/Mage/Core/Controller/Varien/Action.php(410): Mage_Adminhtml_Controller_Action->preDispatch()
> #12 /app/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(255): Mage_Core_Controller_Varien_Action->dispatch()
> #13 /app/app/code/core/Mage/Core/Controller/Varien/Front.php(181): Mage_Core_Controller_Varien_Router_Standard->match()
> #14 /app/app/code/core/Mage/Core/Model/App.php(358): Mage_Core_Controller_Varien_Front->dispatch()
> #15 /app/app/Mage.php(743): Mage_Core_Model_App->run()
> #16 /app/index.php(71): Mage::run()
> #17 {main}
> 

Just going in admin to Rights (or privelges?)/roles/resources and save it again seems to have no effect.
I discovered that roles are stored in the table ''admin_rule". Would it be enough to just manually add the rules there? Or would this have unwanted side effects I am not aware of?

I did not post this as a bug, because it seems to come (only) from external plugins (from which I don't expect to get any support from nowadays)

Kind regards

**Comment by addison74** [comment]
Did you clear the cache? Are you using any FPC extension? Redis?

Disable those two extensions and check if the exceptions are written in the log file. Install the file system and database in a test environment. Export the admin_rule table and then truncate it. Check if the errors still appear after it is empty.

**Comment by henrykbrzoska** [comment]
Was it throwed once and only after update?
Got same thing and it is connected with moving Zend from lib/Zend to zf1-future

**Reply by addison74** [reply]
The error shows that the file ```/app/vendor/shardj/zf1-future/library/Zend/Acl.php``` does not exist. Two times.

**Reply by henrykbrzoska** [reply]
error which you added above shows that some Resource not (maby yet) exists.

**Reply by alexh-swdev** [reply]
@henrykbrzoska No I had it way before 20.1
@ADDISON74 No, the file exists. But some resources are missing.

**Reply by henrykbrzoska** [reply]
i found change which make it logging
![image](https://github.com/OpenMage/magento-lts/assets/4395216/32f41a33-30ad-469e-b9da-f17e7a84333b)

exception was throwing all the time but it just wont logged.
commit:
https://github.com/OpenMage/magento-lts/commit/1699468f62ab2881014dadf68938d64de6a270b5#diff-1048571d4baf38d246708d5953af464948b5591c2c8d648036394d6101ba8022
so it was not 'dead code'.

**Reply by henrykbrzoska** [reply]
@daboss84 so basically you can move app/code/core/Mage/Admin/Model/Resource/Acl.php to app/code/local/Mage/Admin/Model/Resource/Acl.php and comment out Mage::logException($e) and it will work like before

**Reply by fballiano** [reply]
did you check https://github.com/OpenMage/magento-lts/pull/3476?

**Reply by elidrissidev** [reply]
I would recommend just removing the orphan rows from the database as mentioned in #2705 instead of overriding the class.

**Reply by alexh-swdev** [reply]
@henrykbrzoska disable logging is not necessarily a fix (similar to "disconnect the 'check engine' bulp" :) )
However, if this is intended and really can get ignored, it should be noted in the code and not logged at all imho...

**Reply by fballiano** [reply]
my patch didn't disable logging, but maybe I'm commenting in the wrong thread 😅

**Reply by henrykbrzoska** [reply]
But there wasn't throwed exception all the time before. Logging every mess to exception.log will make it unreadable(similar to disco on car dashboard bulbs :D)

[elidrissidev](https://github.com/elidrissidev) and what if i turn off some extension and reenable it? will ACL work properly? I don't think so...

**Comment by alexh-swdev** [comment]
I did some testing with the resource "admin/rma/settings/action":
- When chosing "all" resources from the dropdown, this one is missing in the resources-table.
- When chosing "userdefined" and check all resources manually, it is added to the table

Is there some kind of event some plug ins probably don't hook correctly into? Is there a good location for a breakpoint for debugging this saving? I'll investigate further but don't know if I find enough time next week.

**Reply by henrykbrzoska** [reply]
that's clear, because when you are 'all', you are superadmin and you not use Zend Acl

**Reply by alexh-swdev** [reply]
I experience this Exception when logging in with a user whose role has "all" (from the dropdown) resources.

**Reply by alexh-swdev** [reply]
So, if only one rule ("all" for the admin role) is in the admin_rule table, there is no exception:
 
![grafik](https://github.com/OpenMage/magento-lts/assets/18015530/d73f859c-c746-417a-bb44-eb282cfaeb4e)

If there is any other rule for any other role, these exceptions return. Even if I login with the admin user with the admin role having onlyl "all" in this table.
Question is: Why? I don't get it....

**Comment by mbattistini** [comment]
Same problem here, on a test server fortunately, any solutions?

**Comment by kiatng** [comment]
Duplicate of discussion #2705, which is fixed in PR #2706. For 3rd-party rules, you can just remove them in your database:

```sql
DELETE FROM admin_rule 
WHERE `resource_id` IN (
  'admin/rma/settings/action',
  'admin/Symmetrics_CashTicket/action', 
# ... and so on
)
```

**Reply by alexh-swdev** [reply]
Removing did not change anything

**Comment by kiatng** [comment]
Sorry, reopen because the errors may not be related to #2705.

**Comment by S0FTWEX** [comment]
To get rid of these errors in log file just follow these steps:

1. Go to `System -> Permissions -> Roles`
2. Open each role and click on `Role Resources` tab to see rules tree.
3. Now click `Save Role` with no change in it (just confirm Current Admin Password in `Role Info` tab)
4. Done! The table `admin_rule` is now regenerated with actual rules and log errors are gone!

Please, let me know if it's working for you.

**Reply by alexh-swdev** [reply]
No, it's not.
(Sorry, I already wrote/imlied, thatI did this: https://github.com/OpenMage/magento-lts/discussions/3468#discussioncomment-6866305 )

**Reply by kiatng** [reply]
> I did not post this as a bug, because it seems to come (only) from external plugins (from which I don't expect to get any support from nowadays)

@daboss84 Is it possible it is caused by a 3rd-party extension?

**Reply by alexh-swdev** [reply]
It (most) definetly is :) (That's why I did not post it as a bug - as I wrote in the OP :) )
But I don't really have an idea where to look in the code to takle this. What I know so far:

- My idea form my OP: "adding the missing rules manually" is not necessary, they are in the table if not "all" is selected
- In my understanding, "all" should always allow everything
- When I have more than one role, and this other role does not have "all", I get these exceptions, even when logging in with the "all" role

Especially the last point really puzzles me, currently.

For now I just removed all roles except my admin role becauset they are not really in use right now. (where in the past, though)

**Comment by S0FTWEX** [comment]
> It (most) definetly is :) (That's why I did not post it as a bug - as I wrote in the OP :) ) But I don't really have an idea where to look in the code to takle this. What I know so far:
> 
>     * My idea form my OP: "adding the missing rules manually" is not necessary, they are in the table if not "all" is selected
> 
>     * In my understanding, "all" should always allow everything
> 
>     * When I have more than one role, and this other role does not have "all", I get these exceptions, even when logging in with the "all" role
> 
> 
> Especially the last point really puzzles me, currently.
> 
> For now I just removed all roles except my admin role becauset they are not really in use right now. (where in the past, though)

Well the problem is not in admin_rule table, but in the code.
Please, open any of these files in your Symmetrics_CashTicket extension to find and handle the string 'admin/Symmetrics_CashTicket/action':
`etc/config.xml`, `etc/adminhtml.xml`, `etc/acl.xml`
Find this non-existent action (in XML tree `admin -> Symmetrics_CashTicket -> action`) and remove it.

Do the same thing in RMA extension.

**Reply by alexh-swdev** [reply]
That's one of the first things I did. Those do not appear.
I found the acl-tags (in adminhtml.xml for RMA, e.g.) But I don't really understand them, like I can't connect any of them with 'admin/rma/settings/action' (can I attach files here in the discussion?)

Is there a method which must (or should) get overridden or implemented to validate those checks?

**Reply by S0FTWEX** [reply]
Please, attach the XML files here.

**Reply by alexh-swdev** [reply]
Those are the xml files which define acl's for the rma plugin:
[config - Kopie.txt](https://github.com/OpenMage/magento-lts/files/12576562/config.-.Kopie.txt)
[adminhtml - Kopie.txt](https://github.com/OpenMage/magento-lts/files/12576563/adminhtml.-.Kopie.txt)

**Reply by S0FTWEX** [reply]
These are XML from Mirasvit_MstCore extension, not rma. Attach the rma extension files, and the files from Symmetrics_CashTicket.

**Reply by alexh-swdev** [reply]
Ups, sorry

[config - rma.txt](https://github.com/OpenMage/magento-lts/files/12576896/config.-.rma.txt)
[config - cashticket.txt](https://github.com/OpenMage/magento-lts/files/12576899/config.-.cashticket.txt)

**Comment by S0FTWEX** [comment]
`config - cashticket.txt` is OK, I have commented the problematic piece of code. You can leave it commented or you can remove it.
But `config - rma.txt` doesn't contain `acl` tag. Is there any other file in this extension? (`adminhtml.xml` or `acl.xml`?)

[config-cashticket.txt](https://github.com/OpenMage/magento-lts/files/12579260/config-cashticket.txt)
